File: MidiLink.js
/**
 * MidiLink: JSON-over-MIDI protocol for Bitwig <-> iPhone communication.
 *
 * Channel 15 (status 0x9E): Bitwig -> App
 * Channel 16 (status 0x9F): App -> Bitwig (future)
 *
 * Frame format: 2-byte length header + payload note-ons
 *   Note 0, velocity = length low 7 bits
 *   Note 1, velocity = length high 7 bits
 *   Notes 2..N+1, velocity = ASCII byte value
 *
 * @typedef {Object} BitwigToAppMessage
 * @property {string} [growl] - Growl notification text
 *
 * @typedef {Object} AppToBitwigMessage
 */

class MidiLinkHW {
    static STATUS_TO_APP = 0x9E;   // Channel 15
    static STATUS_TO_BITWIG = 0x9F; // Channel 16

    /**
     * @param {Object} deps
     * @param {Object} deps.midiOutput - MIDI output port with sendMidi(status, data1, data2)
     * @param {Object} [deps.host] - Bitwig host for popup notifications
     * @param {boolean} [deps.debug]
     * @param {Function} [deps.println]
     */
    constructor(deps) {
        deps = deps || {};
        this.midiOutput = deps.midiOutput || null;
        this.host = deps.host || null;
        this.debug = deps.debug || false;
        this.println = deps.println || function() {};
    }

    /**
     * Encode a JS object into an array of {status, data1, data2} MIDI events.
     * @param {BitwigToAppMessage} message
     * @returns {Array<{status: number, data1: number, data2: number}>}
     */
    static encode(message) {
        if (!message) return [];
        var json = JSON.stringify(message);
        var length = json.length;
        if (length > 16383) return []; // 14-bit max

        var events = [];
        // Length header: low 7 bits in note 0, high 7 bits in note 1
        events.push({ status: MidiLinkHW.STATUS_TO_APP, data1: 0, data2: length & 0x7F });
        events.push({ status: MidiLinkHW.STATUS_TO_APP, data1: 1, data2: (length >> 7) & 0x7F });

        // Payload: each char as velocity on notes 2..N+1
        for (var i = 0; i < length; i++) {
            events.push({ status: MidiLinkHW.STATUS_TO_APP, data1: i + 2, data2: json.charCodeAt(i) });
        }
        return events;
    }

    /**
     * Decode an array of {status, data1, data2} MIDI events back to a JS object.
     * @param {Array<{status: number, data1: number, data2: number}>} events
     * @returns {Object|null}
     */
    static decode(events) {
        if (!events || events.length < 2) return null;

        var lengthLow = events[0].data2;
        var lengthHigh = events[1].data2;
        var length = lengthLow | (lengthHigh << 7);

        if (events.length < length + 2) return null;

        var chars = [];
        for (var i = 0; i < length; i++) {
            chars.push(String.fromCharCode(events[i + 2].data2));
        }

        try {
            return JSON.parse(chars.join(''));
        } catch (e) {
            return null;
        }
    }

    /**
     * Send a message to the iPhone app via MIDI.
     * @param {BitwigToAppMessage} message
     */
    send(message) {
        if (!this.midiOutput) return;
        var events = MidiLinkHW.encode(message);
        for (var i = 0; i < events.length; i++) {
            this.midiOutput.sendMidi(events[i].status, events[i].data1, events[i].data2);
        }
        if (this.host) {
            this.host.showPopupNotification("MidiLink: growl sent");
        }
        if (this.debug) this.println("MidiLink: sent " + events.length + " events");
    }
}

var MidiLink = {};
if (typeof module !== 'undefined') module.exports = MidiLinkHW;

File: iphone-drummer/Sources/MidiLink.swift
import Foundation

/// Message sent from Bitwig to the iPhone app via MidiLink protocol.
struct BitwigToAppMessage: Decodable {
    var growl: String?
}

/// Message sent from the iPhone app to Bitwig.
struct AppToBitwigMessage: Codable {
    var growl: String?
}

/// Stateful decoder for MidiLink frames arriving as individual MIDI note-on events.
///
/// Frame format (channel 15, status 0x9E):
///   Note 0, velocity = length low 7 bits
///   Note 1, velocity = length high 7 bits
///   Notes 2..N+1, velocity = ASCII byte value
final class MidiLinkDecoder {
    private var events: [(note: UInt8, velocity: UInt8)] = []
    private var expectedLength: Int?

    /// Feed a single MIDI event. Returns a decoded message when a complete frame is received.
    func feed(status: UInt8, note: UInt8, velocity: UInt8) -> BitwigToAppMessage? {
        // Only accept channel 15 note-on (status 0x9E)
        guard status == 0x9E else { return nil }

        events.append((note: note, velocity: velocity))

        // After 2 events, compute expected length
        if events.count == 2 {
            let low = Int(events[0].velocity)
            let high = Int(events[1].velocity)
            expectedLength = low | (high << 7)
        }

        // Check if frame is complete
        guard let expected = expectedLength, events.count >= expected + 2 else {
            return nil
        }

        // Extract payload bytes
        var chars: [UInt8] = []
        for i in 0..<expected {
            chars.append(events[i + 2].velocity)
        }

        // Reset state
        events.removeAll()
        expectedLength = nil

        // Decode JSON
        let data = Data(chars)
        return try? JSONDecoder().decode(BitwigToAppMessage.self, from: data)
    }

    /// Reset decoder state (e.g., on connection change).
    func reset() {
        events.removeAll()
        expectedLength = nil
    }
}

/// Encodes an AppToBitwigMessage into MidiLink MIDI events (channel 16, status 0x9F).
enum MidiLinkEncoder {
    static func encode(_ message: AppToBitwigMessage) -> [(status: UInt8, note: UInt8, velocity: UInt8)] {
        guard let data = try? JSONEncoder().encode(message) else { return [] }
        let length = data.count
        guard length <= 16383 else { return [] } // 14-bit max

        var events: [(status: UInt8, note: UInt8, velocity: UInt8)] = []
        let status: UInt8 = 0x9F // Channel 16 note-on

        // Length header: low 7 bits in note 0, high 7 bits in note 1
        events.append((status: status, note: 0, velocity: UInt8(length & 0x7F)))
        events.append((status: status, note: 1, velocity: UInt8((length >> 7) & 0x7F)))

        // Payload: each byte as velocity on notes 2..N+1
        for (i, byte) in data.enumerated() {
            events.append((status: status, note: UInt8(i + 2), velocity: byte))
        }
        return events
    }
}
